### PR TITLE
Code improvement

### PR DESCRIPTION
**Pull request by TheZeta**
Long-awaited gui refactoring is finally here.

To divide the main window into separate compartments, I have created five different fxml files namely: main.fxml, inputs.fxml, informative.fxml, map.fxml and empty.fxml. main.fxml contains the other four with fx:include.

Each fxml file has their respective controller class with the naming convention being x.fxml and xController.fxml.

MainController has a reference for each other controller so that changes will apply.

To establish the communication between different controllers, I have deployed Mediator design pattern and created Mediator.java class. This class contains all the methods that are meant to be forwarded to related controller.

Although this is a breakthrough, there are still many features to implement and poorly implemented methods to be refactored as well. Anyway, this version is stable and seems to be working for the cases I have covered.

### CODE REVIEW

**Review by dogukanteber — CHANGES_REQUESTED** [review]
Nice work! Everything is working as planned. I am not sure about some of the things I said so I will be waiting for your response about those.

File: src/main/java/com/targus/ui/controllers/InformativeController.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/controllers/InformativeController.java:18]: I have decided to stick to the JavaFX naming conventions for ui elements. Besides we know that their type is TextField by looking at the definition. While referencing those variables, seeing the txt prefix helps me to quickly recognize them.

File: src/main/java/com/targus/ui/Mediator.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/Mediator.java:14]: Since this project is mediocrely complex, using design patterns but preferring default naming would be more appropriate.

File: src/main/java/com/targus/ui/controllers/InformativeController.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/controllers/InformativeController.java:47]: The justification for that if condition is that display method is invoked pretty much after each action that would cause a change in the optimization problem object such as adding a new sensor or a target. In such cases, there is the possibility that display function will be called even though there are uninitialized variables for displaying. This solution was an answer to that.

File: src/main/java/com/targus/ui/controllers/InformativeController.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/controllers/InformativeController.java:55]: Well, I thought about it, but I wanted to keep InformativeController so simple that it will only contain one method so the overcrowded method. This issue is debatable, but I think the current structure is the factor that refrains InformativeController from adding to its complexity. More methods equal more complexity. Please do not hesitate to share your further thoughts.

File: src/main/java/com/targus/ui/controllers/MapController.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/controllers/MapController.java:35]: This part is still not complete because adding and removing sensors with live fitness value display is still in the process. I will keep your suggestion whilst implementing that feature.

File: src/main/java/com/targus/ui/widgets/Sensor.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/widgets/Sensor.java:9]: This is out of force of habit. I do not initialize static variables in the constructor. I will reconsider doing so.

File: src/main/java/com/targus/ui/widgets/Sensor.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/widgets/Sensor.java:30]: We are moving in the direction of making the user interface as flexible as possible. This flexibility also includes letting the user change the radii and see the changes in the fitness value immediately.

File: src/main/java/com/targus/ui/controllers/EmptyController.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/controllers/EmptyController.java:9]: It was a temporary and phony name. Thanks for reminding me.

File: src/main/java/com/targus/ui/Mediator.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/Mediator.java:44]: In naming those communication methods, I have taken initiative and named them as concisely as possible. But you are right, I will make them more descriptive.

**Comment by TheZeta** [comment]
You have made important remarks but please keep in mind that this whole refactoring phase is still far away from being complete. There are more changes upcoming (css, responsiveness, live monitoring, etc.) and the way I prefer is that, once the application is regarded stable and working correctly, we should merge it into the master branch so that it will be our new baseline. New features will be continuous, and I think it is better if we implement them one step at a time with different branches as to not bloat a single branch.

Nevertheless, I will make the changes you have suggested and update this pull request. Thanks for the solid comments.

File: src/main/java/com/targus/ui/widgets/Sensor.java
Comment (dogukanteber) [reply src/main/java/com/targus/ui/widgets/Sensor.java:9]: It is possible and much more likely to omit initializing these variables. Therefore, we should take precautions as early as possible

File: src/main/java/com/targus/ui/Mediator.java
Comment (dogukanteber) [reply src/main/java/com/targus/ui/Mediator.java:44]: Apart from making the names more descriptive, we can make them more generic by using interfaces. We can have a meeting to talk about it since I do not know much about the implementation details.

**Comment by dogukanteber** [comment]
You are right but I think we should merge every PR by perfecting the code as much as possible. I get your point but since we still have time, we can spend a little extra time following best practices and covering potential problems or bugs so that we do not deal with them later.

We do not have to create another branch after this by the way, by building on top of these changes, though this is totally up to you. 

Thank you for your hard work! If you do not agree with the things I said, feel free to leave your comments so that we find the middle ground to best of the project.

File: src/main/java/com/targus/ui/Mediator.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/Mediator.java:44]: For now, I am planning to keep them as is. I will design an interface that would address your concerns. But for this pull request I think it is okay if I leave them like that. Please can you ignore this one temporarily for this pull request?

Comment (dogukanteber) [reply src/main/java/com/targus/ui/Mediator.java:44]: Sure

**Comment by TheZeta** [comment]
Important update!

The most recently approved pull requested consisted of a change done in Controller.java which does not exist anymore. Since it was a one-line change, I have updated this branch accordingly so that everything is in place.

File: src/main/java/com/targus/ui/widgets/Sensor.java
Comment (dogukanteber) [inline src/main/java/com/targus/ui/widgets/Sensor.java:30]: We can parameterize the magic number. In `Constants` class, we can create `DEFAULT_SENSOR_RANGE` variable and assign this value. You can give another name if you wish.

File: src/main/java/com/targus/ui/widgets/Sensor.java
Comment (dogukanteber) [inline src/main/java/com/targus/ui/widgets/Sensor.java:14]: Why do not we call this `initializeRadius`?

**Comment by dogukanteber** [comment]
Awesome work! I just had two minor comments which are not a big deal. Please tell me what you think about them and make necessary commits if needed. After that, this PR is ready to merge 🚀

**Comment by TheZeta** [comment]
Both of your concerns are valid and were at the back of my head. However in my opinion they do classify as enhancing the code quality whereas this branch here is classified as the foundation of the revolutionized gui code. Just like the ones you have pointed out, this gui code consists of tens of faulty or subpar implementations which are to be addressed once this foundation is laid a.k.a. merged into the master branch. One example is, solve method in InputsController has a terrific implementation and I am aware of it. Again, think of this pull request as the introduction of the skeleton of the renewed version of gui. If those are your only concerns I suggest you to ignore them for now and approve this pull request. Because the version you work on is still using the old gui and any gui-related changes you make (like the one you did in Controller.java) will make me have to take them into account and reflect those changes here as to keep the application running.

File: src/main/java/com/targus/ui/widgets/Sensor.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/widgets/Sensor.java:14]: The initialization of the static variables of Sensor.java turned out to be more tricky than it looks. The solve method in InputsController will be refactored and maybe will result in the creation of new classes. Until then, I left that code there to act as a reminder. Unfortunately the implementation does not happen in a linear fashion but rather in a collective fashion such that some implementations depend on the other implementations being done.

File: src/main/java/com/targus/ui/widgets/Sensor.java
Comment (TheZeta) [reply src/main/java/com/targus/ui/widgets/Sensor.java:30]: You are right, I was thinking of a commit or even a pull request on its own solely focusing on icing the code quality by delegating to constant variables, deleting extra spaces or adding the ones required and optimizing imports. I think it's my mistake not stating those plans under the pull request description. Anyway, this issue here is the perfect candidate for the hotfixes after the approval of this pull request.

**Comment by dogukanteber** [comment]
As of last week, I stopped using the GUI for testing. Therefore, I do not depend on GUI anymore and will not make any changes to it. So, we are good in that part. My concern is that these enhancements will be forgotten if not all, some of them will be forgotten and will cause badly written code. Since this branch does not affect benchmark tests, I do not see any reason why we do not do the necessary changes now.

**Review by dogukanteber — APPROVED** [review]

**Pull request merged**
